parser/auth/sm3.go

```diff
@@ -0,0 +1,384 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash"
+	"strconv"
+)
+
+// The concrete SM3 Cryptographic Hash Algorithm can be accessed in http://www.sca.gov.cn/sca/xwdt/2010-12/17/content_1002389.shtml
+// This implementation of 'type sm3 struct' is modified from https://github.com/tjfoc/gmsm/tree/601ddb090dcf53d7951cc4dcc66276e2b817837c/sm3
+// Some other references:
+// 	https://tools.ietf.org/id/draft-oscca-cfrg-sm3-01.html
+
+/*
+Copyright Suzhou Tongji Fintech Research Institute 2017 All Rights Reserved.
```

Comment:
Can't we import this instead of copying it? Do we want to maintain a SM3 implementation?

Reply:
maybe we can use something like https://github.com/emmansun/gmsm ?

Reply:
SM3 hashing implementation is relatively stable, and almost needs no more improvement in the future. So we can copy the implementation into TiDB.

https://github.com/emmansun/gmsm seems an alternative implementation being maintained with more optimization. If we need acceleration on SM3 hashing in the future, we can use another repo like this.